test/driver_test.dart

```diff
@@ -146,19 +146,18 @@ void main() {
 /// Runs [count] of fake work requests through [driver], and asserts that they
 /// all completed.
 Future _doRequests(
-    {BazelWorkerDriver driver,
-    int count,
-    Function(Future<WorkResponse>) trackWork}) async {
+    {BazelWorkerDriver? driver,
+    int? count,
+    Function(Future<WorkResponse?>)? trackWork}) async {
   // If we create a driver, we need to make sure and terminate it.
   var terminateDriver = driver == null;
   driver ??= BazelWorkerDriver(MockWorker.spawn);
   count ??= 100;
-  terminateDriver ??= true;
   var requests = List.generate(count, (_) => WorkRequest());
   var responses = List.generate(count, (_) => WorkResponse());
   MockWorker.responseQueue.addAll(responses);
   var actualResponses = await Future.wait(
-      requests.map((request) => driver.doWork(request, trackWork: trackWork)));
+      requests.map((request) => driver!.doWork(request, trackWork: trackWork)));
```

Comment:
cc @leafpetersen is it expected that we don't get type promotion here?

Reply:
If there is a write to a variable (anywhere) we undo all promotion of that variable on entry to a closure (since in general we can't track when the closure is called).  For the specific case that all writes happen before the closure is introduced we could do better, but I'm not sure we have the infrastructure for that.  cc @stereotype441

Reply:
> If there is a write to a variable (anywhere) we undo all promotion of that variable on entry to a closure (since in general we can't track when the closure is called). For the specific case that all writes happen before the closure is introduced we could do better, but I'm not sure we have the infrastructure for that. cc @stereotype441

Yeah, the infrastructure for that would be tricky to create, since it would have to happen during the flow analysis "pre pass", which happens during parsing, so the full structure of the user's code isn't very well understood yet.

Reply:
ah ok thanks :)